[FSDP][2/N] Refactor state dict hook registration

awgu opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#90777 [FSDP][2/N] Refactor state dict hook registration**
* #90767 [FSDP][1/N] `fully_shard` state dict
* #90660 [FSDP] Clean up `FlatParamHandle` dtypes, post-backward hook

This PR includes some follow-ups from the previous PR to clean up the code.

Commit:
[FSDP][2/N] Refactor state dict hook registration

[ghstack-poisoned]

Commit:
Update on "[FSDP][2/N] Refactor state dict hook registration"


This PR includes some follow-ups from the previous PR to clean up the code.

[ghstack-poisoned]

rohan-varma commented on torch/distributed/fsdp/_state_dict_utils.py:
```diff
@@ -706,55 +706,40 @@ def _post_load_state_dict_hook(
+    hook_registration_fn_kwargs: Dict[str, Any],
+) -> None:
+    """Registers ``hook`` using ``hook_registration_fn``."""
+    hook_with_state = functools.partial(hook, state)
```
nit: maybe add a TODO and follow up issue to get state directly from module and avoid this partial here. cc @fegin

awgu replied:
Done.

rohan-varma reviewed (approved):
LGTM, thanks for addressing the feedback!

Commit:
Update on "[FSDP][2/N] Refactor state dict hook registration"


This PR includes some follow-ups from the previous PR to clean up the code.

[ghstack-poisoned]

awgu commented:
@pytorchbot merge
